k8sutils/pkg/cri/criwrapper.go

```diff
@@ -0,0 +1,180 @@
+package criwrapper
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	odigosv1 "github.com/odigos-io/odigos/api/odigos/v1alpha1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	criapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+type CriClient struct {
+	conn   *grpc.ClientConn
+	client criapi.RuntimeServiceClient
+	Logger logr.Logger
+}
+
+// Default runtime endpoints
+var defaultRuntimeEndpoints = []string{
+	"unix:///run/containerd/containerd.sock",
+	"unix:///run/crio/crio.sock",
+	"unix:///var/run/cri-dockerd.sock",
+}
+
+func detectRuntimeSocket() string {
+	for _, endpoint := range defaultRuntimeEndpoints {
+		// Extract the file path from the endpoint
+		socketPath := strings.TrimPrefix(endpoint, "unix://")
+		if _, err := os.Stat(socketPath); err == nil {
+			return endpoint
+		}
+	}
+	return ""
+}
+
+// Connect attempts to establish a connection to a CRI runtime.
+func (rc *CriClient) Connect(ctx context.Context) error {
+	var err error
+
+	endpoint := detectRuntimeSocket()
+	if endpoint == "" {
+		return fmt.Errorf("unable to detect CRI runtime endpoint")
+	}
+
+	rc.Logger.Info("Starting connection attempt to CRI runtime", "endpoint", endpoint)
+
+	rc.conn, err = grpc.NewClient(
+		endpoint,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+
+	if err != nil {
+		return fmt.Errorf("failed to create gRPC connection: %v", err)
+	}
+
+	// Create a new RuntimeService client
+	rc.client = criapi.NewRuntimeServiceClient(rc.conn)
+
+	// Validate the connection by invoking a lightweight method
+	ctx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
+
+	_, err = rc.client.Version(ctx, &criapi.VersionRequest{})
+	if err != nil {
+		return fmt.Errorf("Failed to validate CRI runtime connection")
```

Comment:
Same comment as above, I think we can declare this as `ErrFailedToConnectToCRI` and the caller can handle this accordingly